src/ngx_http_lua_api.c

```diff
@@ -74,4 +86,121 @@ ngx_http_lua_add_package_preload(ngx_conf_t *cf, const char *package,
     return NGX_OK;
 }
 
+
+ngx_shm_zone_t *
+ngx_http_lua_shared_memory_add(ngx_conf_t *cf, ngx_str_t *name, size_t size,
+                               void *tag)
+{
+    ngx_http_lua_main_conf_t    *lmcf;
+    ngx_shm_zone_t              **zp;
+    ngx_shm_zone_t              *zone;
+    ngx_http_lua_shm_zone_ctx_t *ctx;
+    ngx_int_t                   n;
+
+    lmcf = ngx_http_conf_get_module_main_conf(cf, ngx_http_lua_module);
+    if (lmcf == NULL) {
+        return NULL;
+    }
+
+    if (lmcf->shm_zones == NULL) {
+        lmcf->shm_zones = ngx_palloc(cf->pool, sizeof(ngx_array_t));
+        if (lmcf->shm_zones == NULL) {
+            return NULL;
+        }
+
+        if (ngx_array_init(lmcf->shm_zones, cf->pool, 2,
+                           sizeof(ngx_shm_zone_t *))
+            != NGX_OK)
+        {
+            return NULL;
+        }
+    }
+
+    zone = ngx_shared_memory_add(cf, name, (size_t) size, tag);
+    if (zone == NULL) {
+        return NULL;
+    }
+
+    if (zone->data) {
+        ctx = (ngx_http_lua_shm_zone_ctx_t *)zone->data;
+        return (ngx_shm_zone_t *)&ctx->data;
+    }
+
+    n = offsetof(ngx_http_lua_shm_zone_ctx_t, data)
+        + sizeof(ngx_shm_zone_t);
+
+    ctx = ngx_pcalloc(cf->pool, n);
+    if (ctx == NULL) {
+        return NULL;
+    }
+
+    ctx->lmcf = lmcf;
+    ctx->log = &cf->cycle->new_log;
+    ngx_memcpy(&ctx->data, zone, sizeof(*zone));
+
+    zp = ngx_array_push(lmcf->shm_zones);
+    if (zp == NULL) {
+        return NULL;
+    }
+
+    *zp = zone;
+
+    /* set zone init */
+    zone->init = ngx_http_lua_shared_memory_init;
+    zone->data = ctx;
+
+    lmcf->requires_shm = 1;
+
+    return (ngx_shm_zone_t *)&ctx->data;
+}
+
+
+static ngx_int_t
+ngx_http_lua_shared_memory_init(ngx_shm_zone_t *shm_zone, void *data)
+{
+    ngx_http_lua_shm_zone_ctx_t *octx = data;
+    ngx_shm_zone_t              *ozone;
+    void                        *odata;
+
+    ngx_http_lua_main_conf_t    *lmcf;
+    ngx_http_lua_shm_zone_ctx_t *ctx;
+    ngx_shm_zone_t              *zone;
+
+    ctx = (ngx_http_lua_shm_zone_ctx_t *)shm_zone->data;
+    zone = (ngx_shm_zone_t *)&ctx->data;
+
+    odata = NULL;
+    if (octx) {
+        ozone = (ngx_shm_zone_t *)&octx->data;
+        odata = ozone->data;
+        zone->shm  = ozone->shm;
+        zone->noreuse = ozone->noreuse;
+
+    } else {
+        zone->shm = shm_zone->shm;
+        zone->noreuse = shm_zone->noreuse;
+    }
+
+    if (zone->init(zone, odata) != NGX_OK) {
+        return NGX_ERROR;
+    }
+
+    lmcf = ctx->lmcf;
+    if (lmcf == NULL) {
+        return NGX_ERROR;
+    }
+
+    lmcf->shm_zones_inited++;
+
+    if (lmcf->shm_zones_inited == lmcf->shm_zones->nelts
+        && lmcf->init_handler)
+    {
+        if (lmcf->init_handler(ctx->log, lmcf, lmcf->lua) != NGX_OK) {
+            /* an error happened */
+            return NGX_ERROR;
+        }
+    }
+
+    return NGX_OK;
+}
```

Comment:
`ngx_http_lua_shared_memory_init` ensure all the shm zones are properly initialized before the Lua code in init_by_lua runs.